runtimevar/runtimevar.go

```diff
@@ -42,32 +43,76 @@ type Snapshot struct {
 
 // Variable provides the ability to read runtime variables with its blocking Watch method.
 type Variable struct {
-	watcher  driver.Watcher
-	nextCall time.Time
-	prev     driver.State
+	watcher    driver.Watcher
+	nextCall   time.Time
+	prev       driver.State
+	ctx        context.Context
+	cancel     func()
+	initLatest sync.Once
+
+	mu       sync.Mutex
+	lastGood interface{}
 }
 
 // New constructs a Variable object given a driver.Watcher implementation.
 func New(w driver.Watcher) *Variable {
-	return &Variable{watcher: w}
+	// ctx/cancel are used to manage the background goroutine in Latest.
+	ctx, cancel := context.WithCancel(context.Background())
+	return &Variable{
+		watcher: w,
+		ctx:     ctx,
+		cancel:  cancel,
+	}
 }
 
-// Watch blocks until there are variable changes, the Context's Done channel
-// closes or a new error occurs.
+// Latest returns the latest successfully retrieved value of the variable.
+//
+// The first call to Latest will block until the variable is successfully
```

Comment:
If this is to block, then it should take in a context for cancellation, which implies having an error.

The more complex (but correct) thing is that what should happen here is that the first watch should start (but not block) during server startup, but block the server health checks from reporting healthy, and cause any calls to `Latest` to immediately fail without blocking (since a server can still receive traffic while in unhealthy).

Reply:
Updated to take a context and return an error.